### PR TITLE
Create WHILE operator, update IF operator

### PR DESCRIPTION
**Pull request by njeffrie**
Create WHILE operator and split logic to copy into and out of subgraphs into kernel_util.cc/.h. Update IF operator to use these methods as well.

Fix subtle bug in micro_allocation_info.cc where unused outputs had the incorrect lifetime (it was previously assumed that all outputs would be used, but not all models meet this invariant).

BUG=http://b/223194691

Mirrors http://cl/432961452

### CODE REVIEW

File: tensorflow/lite/micro/kernels/Makefile.inc [hunk 1]
```diff
@@ -102,6 +101,7 @@ tensorflow/lite/micro/kernels/svdf_test.cc \
 tensorflow/lite/micro/kernels/tanh_test.cc \
 tensorflow/lite/micro/kernels/transpose_test.cc \
 tensorflow/lite/micro/kernels/unpack_test.cc \
+tensorflow/lite/micro/kernels/while_test.cc \
```
Comment (deqiangc) [inline tensorflow/lite/micro/kernels/Makefile.inc:104]: Merge error:? broadcast_to_test.cc is removed

File: tensorflow/lite/micro/tools/make/Makefile [hunk 2]
```diff
@@ -382,6 +381,7 @@ tensorflow/lite/micro/kernels/transpose.cc \
 tensorflow/lite/micro/kernels/transpose_conv.cc \
 tensorflow/lite/micro/kernels/unpack.cc \
 tensorflow/lite/micro/kernels/var_handle.cc \
+tensorflow/lite/micro/kernels/while.cc \
```
Comment (deqiangc) [inline tensorflow/lite/micro/tools/make/Makefile:384]: merge error:?

File: tensorflow/lite/micro/kernels/BUILD [hunk 3]
```diff
@@ -139,8 +139,6 @@ tflm_kernel_cc_library(
         "arg_min_max.cc",
         "assign_variable.cc",
         "batch_to_space_nd.cc",
-        "broadcast_args.cc",
```
Comment (deqiangc) [inline tensorflow/lite/micro/kernels/BUILD:142]: merge error

File: tensorflow/lite/micro/kernels/BUILD [hunk 4]
```diff
@@ -373,34 +372,6 @@ cc_test(
     ],
 )
 
-cc_test(
-    name = "broadcast_args_test",
-    srcs = [
```
Comment (deqiangc) [inline tensorflow/lite/micro/kernels/BUILD:378]: merge error

**Review by deqiangc — CHANGES_REQUESTED** [review]
Pls fix some merge errors

File: tensorflow/lite/micro/kernels/Makefile.inc (hunk 1 above)
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/Makefile.inc:105]: Fixed.

File: tensorflow/lite/micro/tools/make/Makefile (hunk 2 above)
Comment (njeffrie) [reply tensorflow/lite/micro/tools/make/Makefile:385]: Fixed.

File: tensorflow/lite/micro/kernels/while.cc [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <stddef.h>
+
+#include <cstring>
```
Comment (deqiangc) [inline tensorflow/lite/micro/kernels/while.cc:18]: sorry, nitpicking: is cstring really needed here?

**Pull request merged**